clang/lib/Sema/HLSLExternalSemaSource.cpp

```diff
@@ -674,6 +700,34 @@ BuiltinTypeDeclBuilder::addHandleAccessFunction(DeclarationName &Name,
       .finalizeMethod();
 }
 
+BuiltinTypeDeclBuilder &BuiltinTypeDeclBuilder::addAppendMethod() {
+  using PH = BuiltinTypeMethodBuilder::PlaceHolder;
+  ASTContext &AST = SemaRef.getASTContext();
+  QualType ElemTy = getHandleElementType();
+  return BuiltinTypeMethodBuilder(*this, "Append", AST.VoidTy)
+      .addParam("value", ElemTy)
+      .callBuiltin("__builtin_hlsl_buffer_update_counter", AST.UnsignedIntTy,
+                   PH::Handle, getConstantIntExpr(1))
+      .callBuiltin("__builtin_hlsl_resource_getpointer",
+                   AST.getPointerType(ElemTy), PH::Handle, PH::LastStmtPop)
+      .dereference()
+      .assign(PH::_0)
```

Comment:
Should we make all of the methods that operate on the last statement use this explicitly? Something like:
```suggestion
      .callBuiltin("__builtin_hlsl_resource_getpointer",
                   AST.getPointerType(ElemTy), PH::Handle, PH::LastStmtPop)
      .dereference(PH::LastStmtPop)
      .assign(PH::LastStmtPop, PH::_0)
```

I'm not sure if this is better or worse though.

I also think in that case we could maybe just shorten this to `LastStmt` since we'd be using it a lot.

Reply:
I think that's a good idea, it makes it clear what gets dereferenced or assigned. I'll change the name to `LastStmt` because we probably always need to pop it.